### ISSUE TITLE
Should I expect a fully portable output?

### ISSUE THREAD

**Issue by linluc** [Nuitka/Nuitka#958]
Hi, here is my use case for Nuitka, which may not be typical. 

I have laptop A running Ubuntu 16 and python 2.7. On that laptop I have a python2/QT4 application installed and running which I would like to preserve for as long as I can.  My another laptop B is running Ubuntu 20 (no python2 available, and python3 does not have official bindings for QT4). For those reasons I cannot simply run the program (I am not the author, but I have the python2 source) on laptop B. My idea is to recompile the application on laptop A and then run it on laptop B as the binary.

After many, many attempts I made it compile successfully with Nuitka (gstreamer was the most problematic part), it works perfectly on laptop A. 

Now I am trying to run it on laptop B by simply copying the "dist" directory. First I get a series of warnings about fonts, which for now I can ignore. Then I get many warnings about gstreamer libraries not loaded properly. The application starts, but it cannot play any sound, which is one of the most important features.

I went back to my laptop A, ran the compiled binary and checked "lsof -p PID" to see what libs the program has loaded. Surprisingly, the program is still loading many libraries from "/usr/lib", also many gstreamer libraries are loaded from there. Although some gstreamer files are loaded from the "dist" directory where the binary sits. I suspect this is the reason why gstreamer cannot load properly on laptop B. Anyway, I am stuck now, my Nuitka version:

`0.6.11.1
Python: 2.7.12 (default, Oct  5 2020, 13:56:01) 
Executable: /usr/bin/python
OS: Linux
Arch: x86_64`

I start it as:

`python -m nuitka application.py --show-progress --standalone --plugin-enable=qt-plugins --include-qt-plugins=iconengines,imageformats,phonon_backend`

I have also played with PyInstaller which I noticed copies ALL gstreamer files and plugins to the "dist" directory using a dedicated GST hook. But it fails to start (seg. fault) so it was a dead end, after many hours spent. Based on that  I have tried to force Nuitka to include gstreamer plugins by adding this parameter:

`--include-plugin-directory=/usr/lib/x86_64-linux-gnu/gstreamer-1.0/`

But it had no effect (the size of "dist" dir was the same with or without that parameter). I did not find any examples on how to use this switch, so maybe I used it incorrectly.

I would appreciate some tips or suggestions how I could make Nuitka work better for this special need of mine. Or maybe my vision of a truly portable Linux binary app is utopian to begin with?

**Comment by kayhayen** [comment]
The gstreamer might need some help to know where to load its plugins from, something which is not in Nuitka right now. When this options is talking about plugins, this is for Python files, not C files.

If you were to analyse what the hook of PyInstaller does and specify that, it would be needed.

Nuitka is taking care to make binaries run on other systems in ways that others don't attempt to. But I am not just going to find out about every library there is. For commercial rates, I could see myself doing something. Otherwise I need information on a platter.

**Comment by linluc** [comment]
Thanks, understood. The problematic situation is that when running the compiled version of my application some of the Gstreamer libraries (libgst*) are loaded from the "dist" directory, and some from "/usr/lib". So while it works OK on my laptop A, it causes a compatibility conflict on laptop B. I don't think there is an easy way to avoid that except doing what PyInstaller is doing: create a custom code/hook dedicated to Gstreamer, trace all its libs, grab them and include in the dist folder. But even that is not easy, as PyInstaller created binary is giving me segmentation fault. And of course Gstreamer is not the only soft that will require custom handling - PyInstaller has hundreds of custom hooks. I also played with two env variables for Gstreamer: GST_PLUGIN_SCANNER and GST_PLUGIN_PATH. But this led to new problems on laptop B, with more audio libraries conflicting.

For now I used the opposite approach: I completely removed audio and Gstreamer dependencies form the python code and introduced a new config variable in the application: shell sound player. So I play all sounds as a basic os call: os.system(player, mp3_file).

This works for my needs and now both Nuitka and PyInstaller compile successfully. 

Thanks for this great piece of code you created!
